### PR TITLE
feat: use trace-context "headers" in SNS and SQS single event message attributes for lambda transaction

### PR DESCRIPTION
**Pull request by trentm**
- If a SNS or SQS single event trigger to an instrumented Lambda
  function includes message attributes with the name "traceparent" (and
  "tracestate"), case-insensitive, then those are used to continue the
  trace. This was already being done for API Gateway event headers.
- Also, fix a bug in the capturing of SNS and SQS message attributes as
  "transaction.context.message.headers". Note that only "String"-type
  message attributes are captured. Binary-type attributes are base64
  encoded. I'm not sure of the value in capturing them.

Closes: #1831
Fixes: #2605
Refs: https://github.com/elastic/apm/pull/614

### Checklist

- [x] Implement code
- [x] Add tests
- [x] Add CHANGELOG.asciidoc entry

### CODE REVIEW

File: lib/lambda.js
Comment (astorm) [inline lib/lambda.js:14]: I see we've replaced the smaller functions that identify an individual payload type (ex. `isApiGatewayEvent`) with this larger/more-complex function `triggerTypeFromEvent` function that identifies an individual payload type, and then we call `triggerTypeFromEvent` once its result value is passed around where it needs to be.  The main tradeoffs here seem to be 

- the perf. of multiple calls to the older small function vs. the higher memory use of keeping the constant-ish `TRIGGER_` variables around. 

- Swapping where the complexity goes :) 

No objections, just noting the change out loud.

Comment (astorm) [inline lib/lambda.js:14]: I see we've replaced the smaller functions that identify an individual payload type (ex. `isApiGatewayEvent`) with this larger/more-complex function `triggerTypeFromEvent` function that identifies an individual payload type, and then we call `triggerTypeFromEvent` once its result value is passed around where it needs to be.  The main tradeoffs here seem to be 

- the perf. of multiple calls to the older small function vs. the higher memory use of keeping the constant-ish `TRIGGER_` variables around. 

- Swapping where the complexity goes :) 

No objections, just noting the change out loud.

File: lib/lambda.js
Comment (astorm) [inline lib/lambda.js:225]: :+1: for `Object.keys(obj)` -- I'm pretty sure that's the fastest way to do this without needing to get into `for in` and `hasOwnProperty` territory. Unrelated to this PR/note-to-self -- a custom eslint linting rule that ensures we're using the more efficient Object.keys() for object looping/iteration (vs. one of the other myriad ways you can loop an object in javascript) would be a nice to have.

Comment (astorm) [inline lib/lambda.js:225]: :+1: for `Object.keys(obj)` -- I'm pretty sure that's the fastest way to do this without needing to get into `for in` and `hasOwnProperty` territory. Unrelated to this PR/note-to-self -- a custom eslint linting rule that ensures we're using the more efficient Object.keys() for object looping/iteration (vs. one of the other myriad ways you can loop an object in javascript) would be a nice to have.

File: lib/lambda.js
Comment (astorm) [inline lib/lambda.js:613]: 👍 I see a new `headersFromSqsMessageAttributes` that simplifies the data from amazon's strong-ishly typed JSON data into something that APM Server can handle. (I also see a separate function for SNS messages)

Comment (astorm) [inline lib/lambda.js:613]: 👍 I see a new `headersFromSqsMessageAttributes` that simplifies the data from amazon's strong-ishly typed JSON data into something that APM Server can handle. (I also see a separate function for SNS messages)

File: lib/lambda.js
Comment (astorm) [inline lib/lambda.js:413]: 👍  for guarding against unknown datatypes here, per previous discussions on how things can go wrong without this.

Comment (astorm) [inline lib/lambda.js:413]: 👍  for guarding against unknown datatypes here, per previous discussions on how things can go wrong without this.

File: lib/lambda.js
Comment (astorm) [inline lib/lambda.js:459]: Not required for this PR, but I'm genuinely unsure if we should mention the agent's  automatic handling of reading the trace context headers in a Lambda environment in our message queue doc, or if that would confusing things more https://github.com/elastic/apm-agent-nodejs/blob/main/docs/message-queues.asciidoc

Comment (astorm) [inline lib/lambda.js:459]: Not required for this PR, but I'm genuinely unsure if we should mention the agent's  automatic handling of reading the trace context headers in a Lambda environment in our message queue doc, or if that would confusing things more https://github.com/elastic/apm-agent-nodejs/blob/main/docs/message-queues.asciidoc

**Review by astorm — COMMENTED** [review]
Nothing objectionable here. It looks like _the thing to do_ and the approach seems reasonable.  Happy to approve once a changelog entry gets added.

I noticed this was still in draft state -- if there's a specific part you wanted feedback  on or  a part you think needs more extensive testing from a second set of eyes just let me know.  Otherwise 👍

File: lib/lambda.js
Comment (trentm) [reply lib/lambda.js:459]: I think it is worthwhile to mention this behaviour in the *lambda.asciidoc*. Likely a reference in the message-queues doc would be worthwhile as well, I'd have to re-read that doc.

**Comment by trentm** [comment]
> I noticed this was still in draft state

Ah, thanks. I was going to pull out of draft after the changelog entry and merge from main.

**Comment by trentm** [comment]
Also there is still some discussion on the apm spec: https://github.com/elastic/apm/pull/614#discussion_r823436714
So I'll wait for that to settle before finalizing this.

**Comment by trentm** [comment]
Trying that failed "Integration Tests" again: https://apm-ci.elastic.co/job/apm-integration-tests-selector-mbp/job/main/2446/

**Review by astorm — APPROVED** [review]

**Pull request merged**
